substrate/primitives/runtime/src/generic/unchecked_extrinsic.rs

```diff
@@ -366,106 +549,297 @@ impl<Address: Encode, Signature: Encode, Call: Encode, Extra: SignedExtension> s
 }
 
 #[cfg(feature = "serde")]
-impl<'a, Address: Decode, Signature: Decode, Call: Decode, Extra: SignedExtension>
-	serde::Deserialize<'a> for UncheckedExtrinsic<Address, Call, Signature, Extra>
+impl<'a, Address: Decode, Signature: Decode, Call: Decode, Extension: Decode> serde::Deserialize<'a>
+	for UncheckedExtrinsic<Address, Call, Signature, Extension>
 {
 	fn deserialize<D>(de: D) -> Result<Self, D::Error>
 	where
 		D: serde::Deserializer<'a>,
 	{
 		let r = sp_core::bytes::deserialize(de)?;
-		Decode::decode(&mut &r[..])
+		Self::decode(&mut &r[..])
 			.map_err(|e| serde::de::Error::custom(format!("Decode error: {}", e)))
 	}
 }
 
-impl<Address, Call, Signature, Extra> fmt::Debug
-	for UncheckedExtrinsic<Address, Call, Signature, Extra>
+/// A payload that has been signed for an unchecked extrinsics.
+///
+/// Note that the payload that we sign to produce unchecked extrinsic signature
+/// is going to be different than the `SignaturePayload` - so the thing the extrinsic
+/// actually contains.
+pub struct SignedPayload<Call: Dispatchable, Extension: TransactionExtension<Call>>(
+	(Call, Extension, Extension::Implicit),
+	ExtrinsicVersion,
+);
+
+impl<Call, Extension> SignedPayload<Call, Extension>
 where
-	Address: fmt::Debug,
-	Call: fmt::Debug,
-	Extra: SignedExtension,
+	Call: Encode + Dispatchable,
+	Extension: TransactionExtension<Call>,
 {
-	fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
-		write!(
-			f,
-			"UncheckedExtrinsic({:?}, {:?})",
-			self.signature.as_ref().map(|x| (&x.0, &x.2)),
-			self.function,
-		)
+	/// Create new `SignedPayload`.
+	///
+	/// This function may fail if `implicit` of `Extension` is not available.
+	pub fn new(call: Call, tx_ext: Extension) -> Result<Self, TransactionValidityError> {
+		let implicit = Extension::implicit(&tx_ext)?;
+		let raw_payload = (call, tx_ext, implicit);
+		Ok(Self(raw_payload, EXTRINSIC_FORMAT_VERSION))
+	}
+
+	/// Create new legacy `SignedPayload` for extrinsic format version 4.
+	///
+	/// This function may fail if `implicit` of `Extension` is not available.
+	pub fn new_legacy(call: Call, tx_ext: Extension) -> Result<Self, TransactionValidityError> {
+		let implicit = Extension::implicit(&tx_ext)?;
+		let raw_payload = (call, tx_ext, implicit);
+		Ok(Self(raw_payload, LOWEST_SUPPORTED_EXTRINSIC_FORMAT_VERSION))
+	}
+
+	/// Create new `SignedPayload` from raw components.
+	pub fn from_raw(call: Call, tx_ext: Extension, implicit: Extension::Implicit) -> Self {
+		Self((call, tx_ext, implicit), EXTRINSIC_FORMAT_VERSION)
+	}
+
+	/// Deconstruct the payload into it's components.
+	pub fn deconstruct(self) -> (Call, Extension, Extension::Implicit) {
+		self.0
+	}
```

Comment:
The way this type is used is not so ideal to me.

Like here, executing `deconstruct` and then `from_raw` can change the actual value.

And in `Encode` we also lose the extrinsic format version.

Maybe having 2 types: `SignedPayloadV4` and `SignedPayloadV5` would make the code safer.

Reply:
That's intended. We want to just support the old encoding/decoding and disincentivize users from building old extrinsics.

Reply:
yes I just realized this is a new type, so it is not breaking users.